<!DOCTYPE html><html lang="en" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>构建 SAW 分析参考索引 | Louis Base</title><meta name="author" content="Louis Xiong"><meta name="copyright" content="Louis Xiong"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="#ffffff"><meta name="description" content="SAW-ST-V8 make reference.">
<meta property="og:type" content="article">
<meta property="og:title" content="构建 SAW 分析参考索引">
<meta property="og:url" content="http://louis-wemby.github.io/2026/01/11/SAW-makeRef/index.html">
<meta property="og:site_name" content="Louis Base">
<meta property="og:description" content="SAW-ST-V8 make reference.">
<meta property="og:locale" content="en_US">
<meta property="og:image" content="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/29/1/10.1093_bioinformatics_bts635/3/m_bioinformatics_29_1_15_f2.jpeg?Expires=1771021321&Signature=fqt36-Ujsft5yhbtrBM0IVIOgNYIJdnH2nPfHkwb~JZD1RMAQ-Khdm6Qkf6X2m~MbpxQ4D~otvV37upgiKEy7dCA5r5wky4G9PL84IZxKK08t6oBzY1w-LB-gjzFS1gg4RMojlD~WYIgMjkBFkcAWdrrY3ZcoL1NwCkURaodOQbVmyn80P4SD89j3F9ylCvcKp6Dgn3Nj1yLcqgvxu9s-Rgs4cvbsJRWHSxHqpo0JmjtDDyFkzvDXHpwxsMaF60BvwMpvw5hMGBH0D2tBWm7Z~5jOkdpv0QySk~WoALpsNsgJndzXFF3jftYEyqN4h7fyDBN6DJREleM5lyhTwM2ow__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA">
<meta property="article:published_time" content="2026-01-11T16:00:00.000Z">
<meta property="article:modified_time" content="2026-01-11T16:00:00.000Z">
<meta property="article:author" content="Louis Xiong">
<meta property="article:tag" content="BGI">
<meta property="article:tag" content="Genomics">
<meta property="article:tag" content="Spatial Transcriptomics">
<meta property="article:tag" content="Stereo-seq">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/29/1/10.1093_bioinformatics_bts635/3/m_bioinformatics_29_1_15_f2.jpeg?Expires=1771021321&Signature=fqt36-Ujsft5yhbtrBM0IVIOgNYIJdnH2nPfHkwb~JZD1RMAQ-Khdm6Qkf6X2m~MbpxQ4D~otvV37upgiKEy7dCA5r5wky4G9PL84IZxKK08t6oBzY1w-LB-gjzFS1gg4RMojlD~WYIgMjkBFkcAWdrrY3ZcoL1NwCkURaodOQbVmyn80P4SD89j3F9ylCvcKp6Dgn3Nj1yLcqgvxu9s-Rgs4cvbsJRWHSxHqpo0JmjtDDyFkzvDXHpwxsMaF60BvwMpvw5hMGBH0D2tBWm7Z~5jOkdpv0QySk~WoALpsNsgJndzXFF3jftYEyqN4h7fyDBN6DJREleM5lyhTwM2ow__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA"><script type="application/ld+json">{
  "@context": "https://schema.org",
  "@type": "BlogPosting",
  "headline": "构建 SAW 分析参考索引",
  "url": "http://louis-wemby.github.io/2026/01/11/SAW-makeRef/",
  "image": "https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/29/1/10.1093_bioinformatics_bts635/3/m_bioinformatics_29_1_15_f2.jpeg?Expires=1771021321&Signature=fqt36-Ujsft5yhbtrBM0IVIOgNYIJdnH2nPfHkwb~JZD1RMAQ-Khdm6Qkf6X2m~MbpxQ4D~otvV37upgiKEy7dCA5r5wky4G9PL84IZxKK08t6oBzY1w-LB-gjzFS1gg4RMojlD~WYIgMjkBFkcAWdrrY3ZcoL1NwCkURaodOQbVmyn80P4SD89j3F9ylCvcKp6Dgn3Nj1yLcqgvxu9s-Rgs4cvbsJRWHSxHqpo0JmjtDDyFkzvDXHpwxsMaF60BvwMpvw5hMGBH0D2tBWm7Z~5jOkdpv0QySk~WoALpsNsgJndzXFF3jftYEyqN4h7fyDBN6DJREleM5lyhTwM2ow__&Key-Pair-Id=APKAIE5G5CRDK6RD3PGA",
  "datePublished": "2026-01-11T16:00:00.000Z",
  "dateModified": "2026-01-11T16:00:00.000Z",
  "author": [
    {
      "@type": "Person",
      "name": "Louis Xiong",
      "url": "http://louis-wemby.github.io"
    }
  ]
}</script><link rel="shortcut icon" href="/img/favicon.ico"><link rel="canonical" href="http://louis-wemby.github.io/2026/01/11/SAW-makeRef/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css?v=5.5.3"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free@7.1.0/css/all.min.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui@6.1.7/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>
    (() => {
      
    const saveToLocal = {
      set: (key, value, ttl) => {
        if (!ttl) return
        const expiry = Date.now() + ttl * 86400000
        localStorage.setItem(key, JSON.stringify({ value, expiry }))
      },
      get: key => {
        const itemStr = localStorage.getItem(key)
        if (!itemStr) return undefined
        const { value, expiry } = JSON.parse(itemStr)
        if (Date.now() > expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return value
      }
    }

    window.btf = {
      saveToLocal,
      getScript: (url, attr = {}) => new Promise((resolve, reject) => {
        const script = document.createElement('script')
        script.src = url
        script.async = true
        Object.entries(attr).forEach(([key, val]) => script.setAttribute(key, val))
        script.onload = script.onreadystatechange = () => {
          if (!script.readyState || /loaded|complete/.test(script.readyState)) resolve()
        }
        script.onerror = reject
        document.head.appendChild(script)
      }),
      getCSS: (url, id) => new Promise((resolve, reject) => {
        const link = document.createElement('link')
        link.rel = 'stylesheet'
        link.href = url
        if (id) link.id = id
        link.onload = link.onreadystatechange = () => {
          if (!link.readyState || /loaded|complete/.test(link.readyState)) resolve()
        }
        link.onerror = reject
        document.head.appendChild(link)
      }),
      addGlobalFn: (key, fn, name = false, parent = window) => {
        if (!true && key.startsWith('pjax')) return
        const globalFn = parent.globalFn || {}
        globalFn[key] = globalFn[key] || {}
        globalFn[key][name || Object.keys(globalFn[key]).length] = fn
        parent.globalFn = globalFn
      }
    }
  
      
      const activateDarkMode = () => {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      const activateLightMode = () => {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#ffffff')
        }
      }

      btf.activateDarkMode = activateDarkMode
      btf.activateLightMode = activateLightMode

      const theme = saveToLocal.get('theme')
    
          theme === 'dark' ? activateDarkMode() : theme === 'light' ? activateLightMode() : null
        
      
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        document.documentElement.classList.toggle('hide-aside', asideStatus === 'hide')
      }
    
      
    const detectApple = () => {
      if (/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)) {
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
  
    })()
  </script><script>const GLOBAL_CONFIG = {
  root: '/',
  algolia: undefined,
  localSearch: {"path":"/search.xml","preload":false,"top_n_per_article":3,"unescape":false,"pagination":{"enable":false,"hitsPerPage":8},"languages":{"hits_empty":"No results found for: ${query}","hits_stats":"${hits} articles found"}},
  translate: undefined,
  highlight: {"plugin":"highlight.js","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":250,"highlightFullpage":false,"highlightMacStyle":true},
  copy: {
    success: 'Copy Successful',
    error: 'Copy Failed',
    noSupport: 'Browser Not Supported'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just now',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: {"limitCount":250,"languages":{"author":"Author: Louis Xiong","link":"Link: ","source":"Source: Louis Base","info":"Copyright belongs to the author. For commercial use, please contact the author for authorization. For non-commercial use, please indicate the source."}},
  lightbox: 'fancybox',
  Snackbar: {"chs_to_cht":"You have switched to Traditional Chinese","cht_to_chs":"You have switched to Simplified Chinese","day_to_night":"You have switched to Dark Mode","night_to_day":"You have switched to Light Mode","bgLight":"#49b1f5","bgDark":"#1f1f1f","position":"bottom-left"},
  infinitegrid: {
    js: 'https://cdn.jsdelivr.net/npm/@egjs/infinitegrid@4.12.0/dist/infinitegrid.min.js',
    buttonText: 'Load More'
  },
  isPhotoFigcaption: false,
  islazyloadPlugin: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: true,
  },
  autoDarkmode: false
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '构建 SAW 分析参考索引',
  isHighlightShrink: false,
  isToc: true,
  pageType: 'post'
}</script><meta name="generator" content="Hexo 8.1.1"></head><body><div class="bg-animation" id="web_bg"></div><script>(() => {
  const arr = ["background-color: #FF9F1C;","background-color: #8A80E6;","background-color: #EDEAF7;","background-color: #E4C1D9;","background-color: #3A5A60;"]
  const webBgDiv = document.getElementById('web_bg')

  const setRandomBg = () => {
    webBgDiv.style = arr[Math.floor(Math.random() * arr.length)]
    requestAnimationFrame(() => webBgDiv.classList.add('bg-animation'))
  }

  document.addEventListener('pjax:send', () => {
    webBgDiv.style = ''
    webBgDiv.classList.remove('bg-animation')
  })

  document.addEventListener('pjax:complete', setRandomBg)
  document.addEventListener('DOMContentLoaded', setRandomBg)
})()
</script><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img text-center"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="site-data text-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">7</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">3</div></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><span class="site-page group"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></span><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></li><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-image"></i><span> Gallery</span></a></li><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/links/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-dna"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="post-bg" id="page-header" style="background-image: url(https://www.vangoghgallery.com/img/starry_night_full.jpg);"><nav id="nav"><span id="blog-info"><a class="nav-site-title" href="/"><img class="site-icon" src="/img/logo.png" alt="Logo"><span class="site-name">Louis Base</span></a><a class="nav-page-title" href="/"><span class="site-name">构建 SAW 分析参考索引</span><span class="site-name"><i class="fa-solid fa-circle-arrow-left"></i><span>  Back to Home</span></span></a></span><div id="menus"><div id="search-button"><span class="site-page social-icon search"><i class="fas fa-search fa-fw"></i><span> Search</span></span></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><span class="site-page group"><i class="fa-fw fas fa-list"></i><span> List</span><i class="fas fa-chevron-down"></i></span><ul class="menus_item_child"><li><a class="site-page child" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></li><li><a class="site-page child" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></li><li><a class="site-page child" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></li><li><a class="site-page child" href="/gallery/"><i class="fa-fw fas fa-image"></i><span> Gallery</span></a></li><li><a class="site-page child" href="/music/"><i class="fa-fw fas fa-music"></i><span> Music</span></a></li></ul></div><div class="menus_item"><a class="site-page" href="/links/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-dna"></i><span> About</span></a></div></div><div id="toggle-menu"><span class="site-page"><i class="fas fa-bars fa-fw"></i></span></div></div></nav><div id="post-info"><h1 class="post-title">构建 SAW 分析参考索引</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2026-01-11T16:00:00.000Z" title="Created 2026-01-12 00:00:00">2026-01-12</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2026-01-11T16:00:00.000Z" title="Updated 2026-01-12 00:00:00">2026-01-12</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/Tutorial/">Tutorial</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-wordcount"><i class="far fa-file-word fa-fw post-meta-icon"></i><span class="post-meta-label">Word Count:</span><span class="word-count">1.5k</span><span class="post-meta-separator">|</span><i class="far fa-clock fa-fw post-meta-icon"></i><span class="post-meta-label">Reading Time:</span><span>7mins</span></span><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title=""><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post Views:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div></header><main class="layout" id="content-inner"><div id="post"><article class="container post-content" id="article-container"><p>在运行 SAW (Stereo-seq Analysis Workflow) count 分析前，需要先生成参考基因组索引 (index).这一步可以由 <code>SAW makeRef</code> 流程分析实现.我们需要提供的输入文件为参考基因组序列 <code>reference.fna</code> 和对应的注释文件 <code>reference.gff/gtf</code>.</p>
<h2 id="miniprot-基因注释"><a href="#miniprot-基因注释" class="headerlink" title="miniprot 基因注释"></a>miniprot 基因注释</h2><p>由于原先的参考基因组 <code>Larus_argentatus.fna</code> 并未注释出关键的三个视蛋白 (Opsins) 基因 <em>OPN1SW</em>、<em>OPN2SW</em> 与 <em>OPN1LW</em>，因此需要重新对参考基因组进行注释，选用的软件为 <strong>miniprot</strong>，具体注释方法在此不展开论述.最终只注释出位于 Chr1 的 <em>OPN2SW</em> 基因，需要添加的另外两个基因序列所在染色体分别为 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/nuccore/CAYQHM010000238.1">CAYQHM010000238.1.fna</a> 和 <a target="_blank" rel="noopener" href="https://www.ncbi.nlm.nih.gov/nuccore/OZ207386.1/">OZ207386.1.fna</a>，<strong>miniprot</strong> 注释输出的 <code>GFF3</code> 格式文件分别为 <code>CAYQHM010000238.1.gff</code> 和 <code>OZ207386.1.gff</code>.</p>
<h2 id="注释文件格式转换"><a href="#注释文件格式转换" class="headerlink" title="注释文件格式转换"></a>注释文件格式转换</h2><p>原先的基因组注释文件为 <code>.gtf</code> 格式，因此需将 <code>CAYQHM010000238.1.gff</code>和 <code>OZ207386.1.gff</code> 转换为 GTF 格式.所用脚本如下：</p>
<figure class="highlight python"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br><span class="line">30</span><br><span class="line">31</span><br><span class="line">32</span><br><span class="line">33</span><br><span class="line">34</span><br><span class="line">35</span><br><span class="line">36</span><br><span class="line">37</span><br><span class="line">38</span><br><span class="line">39</span><br><span class="line">40</span><br><span class="line">41</span><br><span class="line">42</span><br><span class="line">43</span><br><span class="line">44</span><br><span class="line">45</span><br><span class="line">46</span><br><span class="line">47</span><br><span class="line">48</span><br><span class="line">49</span><br><span class="line">50</span><br><span class="line">51</span><br><span class="line">52</span><br><span class="line">53</span><br><span class="line">54</span><br><span class="line">55</span><br><span class="line">56</span><br><span class="line">57</span><br><span class="line">58</span><br><span class="line">59</span><br><span class="line">60</span><br><span class="line">61</span><br><span class="line">62</span><br><span class="line">63</span><br><span class="line">64</span><br><span class="line">65</span><br><span class="line">66</span><br><span class="line">67</span><br><span class="line">68</span><br><span class="line">69</span><br><span class="line">70</span><br><span class="line">71</span><br><span class="line">72</span><br><span class="line">73</span><br><span class="line">74</span><br><span class="line">75</span><br><span class="line">76</span><br><span class="line">77</span><br><span class="line">78</span><br><span class="line">79</span><br><span class="line">80</span><br><span class="line">81</span><br><span class="line">82</span><br><span class="line">83</span><br><span class="line">84</span><br><span class="line">85</span><br><span class="line">86</span><br><span class="line">87</span><br><span class="line">88</span><br><span class="line">89</span><br><span class="line">90</span><br><span class="line">91</span><br><span class="line">92</span><br><span class="line">93</span><br><span class="line">94</span><br><span class="line">95</span><br><span class="line">96</span><br><span class="line">97</span><br><span class="line">98</span><br><span class="line">99</span><br><span class="line">100</span><br><span class="line">101</span><br><span class="line">102</span><br><span class="line">103</span><br><span class="line">104</span><br><span class="line">105</span><br><span class="line">106</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment">#!/usr/bin/env python3</span></span><br><span class="line"><span class="comment"># -*- coding:utf-8 -*-</span></span><br><span class="line"></span><br><span class="line"><span class="keyword">import</span> sys</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">parse_attributes</span>(<span class="params">attr_str</span>):</span><br><span class="line">    <span class="string">&quot;&quot;&quot;解析 GFF 属性列为字典&quot;&quot;&quot;</span></span><br><span class="line">    attrs = &#123;&#125;</span><br><span class="line">    <span class="keyword">for</span> item <span class="keyword">in</span> attr_str.strip(<span class="string">&#x27;;&#x27;</span>).split(<span class="string">&#x27;;&#x27;</span>):</span><br><span class="line">        <span class="keyword">if</span> <span class="string">&#x27;=&#x27;</span> <span class="keyword">in</span> item:</span><br><span class="line">            k, v = item.split(<span class="string">&#x27;=&#x27;</span>, <span class="number">1</span>)</span><br><span class="line">            attrs[k] = v</span><br><span class="line">        <span class="keyword">elif</span> <span class="string">&#x27; &#x27;</span> <span class="keyword">in</span> item: <span class="comment"># 处理某些类 GTF 格式</span></span><br><span class="line">            parts = item.strip().split(<span class="string">&#x27; &#x27;</span>, <span class="number">1</span>)</span><br><span class="line">            <span class="keyword">if</span> <span class="built_in">len</span>(parts) == <span class="number">2</span>:</span><br><span class="line">                attrs[parts[<span class="number">0</span>]] = parts[<span class="number">1</span>].strip(<span class="string">&#x27;&quot;&#x27;</span>)</span><br><span class="line">    <span class="keyword">return</span> attrs</span><br><span class="line"></span><br><span class="line"><span class="keyword">def</span> <span class="title function_">convert_gff_to_gtf</span>(<span class="params">input_file</span>):</span><br><span class="line">    genes = &#123;&#125; <span class="comment"># 存储基因信息</span></span><br><span class="line">    transcripts = &#123;&#125; <span class="comment"># 存储转录本信息 &#123;rna_id: [feature_lines]&#125;</span></span><br><span class="line"></span><br><span class="line">    <span class="keyword">with</span> <span class="built_in">open</span>(input_file, <span class="string">&#x27;r&#x27;</span>) <span class="keyword">as</span> f:</span><br><span class="line">        <span class="keyword">for</span> line <span class="keyword">in</span> f:</span><br><span class="line">            <span class="keyword">if</span> line.startswith(<span class="string">&#x27;#&#x27;</span>) <span class="keyword">or</span> <span class="keyword">not</span> line.strip():</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">            </span><br><span class="line">            cols = line.strip().split(<span class="string">&#x27;\t&#x27;</span>)</span><br><span class="line">            <span class="keyword">if</span> <span class="built_in">len</span>(cols) &lt; <span class="number">9</span>:</span><br><span class="line">                <span class="keyword">continue</span></span><br><span class="line">                </span><br><span class="line">            feature_type = cols[<span class="number">2</span>]</span><br><span class="line">            attributes = parse_attributes(cols[<span class="number">8</span>])</span><br><span class="line">            </span><br><span class="line">            <span class="keyword">if</span> feature_type == <span class="string">&#x27;mRNA&#x27;</span>:</span><br><span class="line">                rna_id = attributes.get(<span class="string">&#x27;ID&#x27;</span>)</span><br><span class="line">                <span class="comment"># 如果没有 Parent，则用 rna_id 作为 gene_id</span></span><br><span class="line">                gene_id = attributes.get(<span class="string">&#x27;Parent&#x27;</span>, rna_id)</span><br><span class="line">                transcripts[rna_id] = &#123;</span><br><span class="line">                    <span class="string">&#x27;info&#x27;</span>: cols,</span><br><span class="line">                    <span class="string">&#x27;gene_id&#x27;</span>: gene_id,</span><br><span class="line">                    <span class="string">&#x27;features&#x27;</span>: []</span><br><span class="line">                &#125;</span><br><span class="line">            <span class="keyword">elif</span> feature_type <span class="keyword">in</span> [<span class="string">&#x27;CDS&#x27;</span>, <span class="string">&#x27;exon&#x27;</span>, <span class="string">&#x27;five_prime_utr&#x27;</span>, <span class="string">&#x27;three_prime_utr&#x27;</span>]:</span><br><span class="line">                rna_id = attributes.get(<span class="string">&#x27;Parent&#x27;</span>)</span><br><span class="line">                <span class="keyword">if</span> rna_id <span class="keyword">in</span> transcripts:</span><br><span class="line">                    transcripts[rna_id][<span class="string">&#x27;features&#x27;</span>].append(cols)</span><br><span class="line"></span><br><span class="line">    <span class="comment"># 开始输出</span></span><br><span class="line">    <span class="keyword">for</span> rna_id, data <span class="keyword">in</span> transcripts.items():</span><br><span class="line">        t_info = data[<span class="string">&#x27;info&#x27;</span>]</span><br><span class="line">        gene_id = data[<span class="string">&#x27;gene_id&#x27;</span>]</span><br><span class="line">        chrom, source, _, start, end, score, strand, frame = t_info[:<span class="number">8</span>]</span><br><span class="line"></span><br><span class="line">        <span class="comment"># 1. 输出 gene 行</span></span><br><span class="line">        gene_attr = <span class="string">f&#x27;gene_id &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;; gene_name &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;; gene_biotype &quot;protein-coding&quot;;&#x27;</span></span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\tgene\t<span class="subst">&#123;start&#125;</span>\t<span class="subst">&#123;end&#125;</span>\t<span class="subst">&#123;score&#125;</span>\t<span class="subst">&#123;strand&#125;</span>\t.\t<span class="subst">&#123;gene_attr&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line">        <span class="comment"># 2. 输出 transcript 行</span></span><br><span class="line">        tran_attr = <span class="string">f&#x27;gene_id &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;; transcript_id &quot;<span class="subst">&#123;rna_id&#125;</span>&quot;; gene_name &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;; gene_biotype &quot;protein-coding&quot;;&#x27;</span></span><br><span class="line">        <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\ttranscript\t<span class="subst">&#123;start&#125;</span>\t<span class="subst">&#123;end&#125;</span>\t<span class="subst">&#123;score&#125;</span>\t<span class="subst">&#123;strand&#125;</span>\t.\t<span class="subst">&#123;tran_attr&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line">        <span class="comment"># 3. 输出子特征 (CDS/Exon)</span></span><br><span class="line">        <span class="comment"># 如果 GFF 里只有 CDS 没有 exon，GTF 必须补全 exon</span></span><br><span class="line">        has_exon = <span class="built_in">any</span>(f[<span class="number">2</span>] == <span class="string">&#x27;exon&#x27;</span> <span class="keyword">for</span> f <span class="keyword">in</span> data[<span class="string">&#x27;features&#x27;</span>])</span><br><span class="line">        </span><br><span class="line">        exon_count = <span class="number">0</span></span><br><span class="line">        cds_count = <span class="number">0</span></span><br><span class="line">        </span><br><span class="line">        <span class="comment"># 排序特征坐标 (正链升序，负链降序)</span></span><br><span class="line">        sorted_features = <span class="built_in">sorted</span>(data[<span class="string">&#x27;features&#x27;</span>], key=<span class="keyword">lambda</span> x: <span class="built_in">int</span>(x[<span class="number">3</span>]))</span><br><span class="line">        </span><br><span class="line">        <span class="keyword">for</span> feat <span class="keyword">in</span> sorted_features:</span><br><span class="line">            f_type = feat[<span class="number">2</span>]</span><br><span class="line">            f_start, f_end, f_score, f_strand, f_frame = feat[<span class="number">3</span>], feat[<span class="number">4</span>], feat[<span class="number">5</span>], feat[<span class="number">6</span>], feat[<span class="number">7</span>]</span><br><span class="line">            </span><br><span class="line">            <span class="comment"># 生成通用属性前缀</span></span><br><span class="line">            base_attr = <span class="string">f&#x27;gene_id &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;; transcript_id &quot;<span class="subst">&#123;rna_id&#125;</span>&quot;; gene_name &quot;<span class="subst">&#123;gene_id&#125;</span>&quot;;&#x27;</span></span><br><span class="line"></span><br><span class="line">            <span class="comment"># 处理 CDS 和 自动生成 exon</span></span><br><span class="line">            <span class="keyword">if</span> f_type == <span class="string">&#x27;CDS&#x27;</span>:</span><br><span class="line">                cds_count += <span class="number">1</span></span><br><span class="line">                <span class="comment"># 如果原始数据没写 exon，把 CDS 当作 exon 输出一次</span></span><br><span class="line">                <span class="keyword">if</span> <span class="keyword">not</span> has_exon:</span><br><span class="line">                    exon_count += <span class="number">1</span></span><br><span class="line">                    ex_attr = base_attr + <span class="string">f&#x27; exon_id &quot;<span class="subst">&#123;rna_id&#125;</span>.exon.<span class="subst">&#123;exon_count&#125;</span>&quot;;&#x27;</span></span><br><span class="line">                    <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\texon\t<span class="subst">&#123;f_start&#125;</span>\t<span class="subst">&#123;f_end&#125;</span>\t<span class="subst">&#123;f_score&#125;</span>\t<span class="subst">&#123;f_strand&#125;</span>\t.\t<span class="subst">&#123;ex_attr&#125;</span>&quot;</span>)</span><br><span class="line">                </span><br><span class="line">                <span class="comment"># 输出 CDS</span></span><br><span class="line">                c_attr = base_attr + <span class="string">f&#x27; exon_id &quot;<span class="subst">&#123;rna_id&#125;</span>.cds.<span class="subst">&#123;cds_count&#125;</span>&quot;;&#x27;</span></span><br><span class="line">                <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\tCDS\t<span class="subst">&#123;f_start&#125;</span>\t<span class="subst">&#123;f_end&#125;</span>\t<span class="subst">&#123;f_score&#125;</span>\t<span class="subst">&#123;f_strand&#125;</span>\t<span class="subst">&#123;f_frame&#125;</span>\t<span class="subst">&#123;c_attr&#125;</span>&quot;</span>)</span><br><span class="line">            </span><br><span class="line">            <span class="keyword">elif</span> f_type == <span class="string">&#x27;exon&#x27;</span>:</span><br><span class="line">                exon_count += <span class="number">1</span></span><br><span class="line">                ex_attr = base_attr + <span class="string">f&#x27; exon_id &quot;<span class="subst">&#123;rna_id&#125;</span>.exon.<span class="subst">&#123;exon_count&#125;</span>&quot;;&#x27;</span></span><br><span class="line">                <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\texon\t<span class="subst">&#123;f_start&#125;</span>\t<span class="subst">&#123;f_end&#125;</span>\t<span class="subst">&#123;f_score&#125;</span>\t<span class="subst">&#123;f_strand&#125;</span>\t.\t<span class="subst">&#123;ex_attr&#125;</span>&quot;</span>)</span><br><span class="line">            </span><br><span class="line">            <span class="keyword">else</span>:</span><br><span class="line">                <span class="comment"># 处理 UTR 等其他特征</span></span><br><span class="line">                <span class="built_in">print</span>(<span class="string">f&quot;<span class="subst">&#123;chrom&#125;</span>\t<span class="subst">&#123;source&#125;</span>\t<span class="subst">&#123;f_type&#125;</span>\t<span class="subst">&#123;f_start&#125;</span>\t<span class="subst">&#123;f_end&#125;</span>\t<span class="subst">&#123;f_score&#125;</span>\t<span class="subst">&#123;f_strand&#125;</span>\t<span class="subst">&#123;f_frame&#125;</span>\t<span class="subst">&#123;base_attr&#125;</span>&quot;</span>)</span><br><span class="line"></span><br><span class="line"><span class="keyword">if</span> __name__ == <span class="string">&#x27;__main__&#x27;</span>:</span><br><span class="line">    <span class="keyword">if</span> <span class="built_in">len</span>(sys.argv) &lt; <span class="number">2</span>:</span><br><span class="line">        <span class="built_in">print</span>(<span class="string">&quot;Usage: python gff2gtf.py input.gff&quot;</span>)</span><br><span class="line">    <span class="keyword">else</span>:</span><br><span class="line">        convert_gff_to_gtf(sys.argv[<span class="number">1</span>])</span><br></pre></td></tr></table></figure>

<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Example</span></span><br><span class="line">python gff2gtf.py CAYQHM010000238.1.gff &gt; OPN1LW.gtf</span><br></pre></td></tr></table></figure>

<p>格式转换成功后，建议将 <code>gene_id</code>、<code>gene_name</code>、<code>transcript_id</code> 等改为基因名（<strong>miniprot</strong> 自动输出的标识为 <code>MPXXXXXX</code>），其余命名格式与原先 GTF 文件相同，如一个转录本（transcript）有多个 exon&#x2F;CDS 时，用 exon.1&#x2F;.2&#x2F;… 和 cds.1&#x2F;.2&#x2F;… 加以区分.</p>
<h2 id="添加序列与注释文件"><a href="#添加序列与注释文件" class="headerlink" title="添加序列与注释文件"></a>添加序列与注释文件</h2><p>在 <strong>Linux</strong> 终端，用如下命令将基因序列与注释信息添加进原有的基因组序列与注释文件中：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># Example</span></span><br><span class="line"><span class="comment"># Chr1 放最前面</span></span><br><span class="line"><span class="built_in">cat</span> OPN2SW.gtf Larus_argentatus.gtf OPN1LW.gtf OPN1SW.gtf &gt; Larus_argentatus_v1.gtf</span><br><span class="line"></span><br><span class="line"><span class="comment"># 删除 &#x27;&gt;&#x27; 后多余信息</span></span><br><span class="line"><span class="built_in">cat</span> Larus_argentatus.fna CAYQHM010000238.1.fna OZ207386.1.fna | sed -e <span class="string">&quot;s/&gt;OZ207386.1 Larus argentatus genome assembly, chromosome: 1/&gt;OZ207386.1/g&quot;</span> -e <span class="string">&quot;s/&gt;CAYQHM010000238.1 Larus argentatus genome assembly, contig: HAP1_SCAFFOLD_368, whole genome shotgun sequence/&gt;CAYQHM010000238.1/g&quot;</span> &gt; Larus_argentatus_v1.fna</span><br></pre></td></tr></table></figure>

<h2 id="运行-SAW-makeRef"><a href="#运行-SAW-makeRef" class="headerlink" title="运行 SAW makeRef"></a>运行 SAW makeRef</h2><p>由于需进行转录组比对，所选用的 Mode 为 <a target="_blank" rel="noopener" href="https://academic.oup.com/bioinformatics/article/29/1/15/272537"><strong>STAR</strong></a>.若在终端运行，则命令如下：</p>
<figure class="highlight bash"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># &#x27;transcript&#x27; 为存放结果文件创建的文件夹</span></span><br><span class="line">saw makeRef \</span><br><span class="line">    --mode=STAR \</span><br><span class="line">    --fasta=Larus_argentatus_v1.fna \</span><br><span class="line">    --gtf=Larus_argentatus_v1.gtf \</span><br><span class="line">    --genome=./transcript</span><br></pre></td></tr></table></figure>

<p>关于 <code>SAW makeRef</code> 更多用法，可见参考文献 1.</p>
<h2 id="⚠️-注意事项"><a href="#⚠️-注意事项" class="headerlink" title="⚠️ 注意事项"></a>⚠️ 注意事项</h2><ol>
<li><code>fna</code> 文件中的染色体&#x2F;Scaffold 等标识符必须唯一，不能重复，且需和 <code>gtf</code> 文件中的对应.</li>
<li>在拼接完基因组序列后，建议检查文件大小&#x2F;行数，防止有碱基序列丢失.可用 <code>wc -l</code> 打印行数，或者 <code>grep -v &#39;&gt;&#39; filename | tr -d &#39;\n&#39; | wc -w</code> 精确统计碱基个数（去除 Header 信息）.</li>
<li>注释文件中的起止坐标不能超过所在染色体&#x2F;Scaffold 的总长度，否则构建索引会报错.</li>
</ol>
<h2 id="参考文献"><a href="#参考文献" class="headerlink" title="参考文献"></a>参考文献</h2><ol>
<li><a target="_blank" rel="noopener" href="https://www.stomics.tech/service/saw_8_2/docs/shi-yong-jiao-cheng/bi-dui-wen-jian/gou-jian-suo-yin-wen-jian.html">构建索引文件</a></li>
<li><a target="_blank" rel="noopener" href="https://academic.oup.com/bioinformatics/article/29/1/15/272537">STAR: ultrafast universal RNA-seq aligner</a></li>
</ol>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta"><i class="fas fa-circle-user fa-fw"></i>Author: </span><span class="post-copyright-info"><a href="http://louis-wemby.github.io">Louis Xiong</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta"><i class="fas fa-square-arrow-up-right fa-fw"></i>Link: </span><span class="post-copyright-info"><a href="http://louis-wemby.github.io/2026/01/11/SAW-makeRef/">http://louis-wemby.github.io/2026/01/11/SAW-makeRef/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta"><i class="fas fa-circle-exclamation fa-fw"></i>Copyright Notice: </span><span class="post-copyright-info">All articles on this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless otherwise stated.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/BGI/">BGI</a><a class="post-meta__tags" href="/tags/Stereo-seq/">Stereo-seq</a><a class="post-meta__tags" href="/tags/Genomics/">Genomics</a><a class="post-meta__tags" href="/tags/Spatial-Transcriptomics/">Spatial Transcriptomics</a></div><div class="post-share"><div class="social-share" data-image="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/29/1/10.1093_bioinformatics_bts635/3/m_bioinformatics_29_1_15_f2.jpeg?Expires=1771021321&amp;Signature=fqt36-Ujsft5yhbtrBM0IVIOgNYIJdnH2nPfHkwb~JZD1RMAQ-Khdm6Qkf6X2m~MbpxQ4D~otvV37upgiKEy7dCA5r5wky4G9PL84IZxKK08t6oBzY1w-LB-gjzFS1gg4RMojlD~WYIgMjkBFkcAWdrrY3ZcoL1NwCkURaodOQbVmyn80P4SD89j3F9ylCvcKp6Dgn3Nj1yLcqgvxu9s-Rgs4cvbsJRWHSxHqpo0JmjtDDyFkzvDXHpwxsMaF60BvwMpvw5hMGBH0D2tBWm7Z~5jOkdpv0QySk~WoALpsNsgJndzXFF3jftYEyqN4h7fyDBN6DJREleM5lyhTwM2ow__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" data-sites="facebook,x,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><a class="pagination-related" href="/2026/01/05/hello-world/" title="Hello World"><img class="cover" src="/figure/posts/hexo.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of previous post"><div class="info"><div class="info-1"><div class="info-item-1">Previous</div><div class="info-item-2">Hello World</div></div><div class="info-2"><div class="info-item-1">Welcome to Hexo! This is your very first post. Check documentation for more info. If you get any problems when using Hexo, you can find the answer in troubleshooting or you can ask me on GitHub. Quick StartCreate a new post1$ hexo new &quot;My New Post&quot;  More info: Writing Run server1$ hexo server  More info: Server Generate static files1$ hexo generate  More info: Generating Deploy to remote sites1$ hexo deploy  More info: Deployment </div></div></div></a><a class="pagination-related" href="/2026/01/12/Lotka-Volterra_model/" title="动态系统稳定性与 Lotka-Volterra 模型浅略研究"><img class="cover" src="https://oaknationalacademy-res.cloudinary.com/image/upload/v1700515161/xvjpxj8bmayxkhjawqkk.png" onerror="onerror=null;src='/img/404.jpg'" alt="cover of next post"><div class="info text-right"><div class="info-1"><div class="info-item-1">Next</div><div class="info-item-2">动态系统稳定性与 Lotka-Volterra 模型浅略研究</div></div><div class="info-2"><div class="info-item-1">稳定性理论（Stability theory）简介稳定性理论研究微分方程的解及动态系统的轨迹在初始条件有微小扰动时的稳定性.在某初始平衡条件（Equilibrium）下，给系统向某一方向施加一微小扰动，若系统最终可能偏离平衡状态而无法回到原始位置，则该平衡点是不稳定的.反之，若为稳定平衡点，则最终能回到原始位置，如摆线的振荡、阻尼振子等系统.           衰减的阻尼正弦波.来自：Wikipedia.   稳定性的判断动态系统可用微分方程来描述.对于一维的常微分方程如 $\frac{dx}{dt} &#x3D; f(x)$，其平衡点 ${x^}$ 满足 $f({x^}) &#x3D; 0$.若 $f’({x^}) &lt; 0$，则平衡点是稳定的；若 $f’({x^*}) &gt; 0$，则平衡点是不稳定的.值得注意的是，若 $f’({x^}) &#x3D; 0$（如下图第二个 stable 处），则需要其他方法判断稳定性.           一维动态系统的稳定性判断.   而对于二维系统 $\frac{dx}{dt} &#x3D; f_1(x_1, x_2)$ 和 $\...</div></div></div></a></nav></div><div class="aside-content" id="aside-content"><div class="card-widget card-info text-center"><div class="avatar-img"><img src="/img/avatar.jpg" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info-name">Louis Xiong</div><div class="author-info-description">A young man who studies bioinformatics and enjoys life.</div><div class="site-data"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">3</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">7</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">3</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/Louis-wemby"><i class="fab fa-github"></i><span>Follow Me</span></a><div class="card-info-social-icons"><a class="social-icon" href="https://github.com/Louis-wemby" target="_blank" title="Github"><i class="fab fa-github" style="color: #24292e;"></i></a><a class="social-icon" href="mailto:didarhossain837463@gmail.com" target="_blank" title="Email"><i class="fas fa-envelope" style="color: #4a7dbe;"></i></a><a class="social-icon" href="https://X.com/Louis_SPURSfan" target="_blank" title="Twitter"><i class="fab fa-x-twitter" style="color: #1da1f2;"></i></a><a class="social-icon" href="https://www.instagram.com/looouis21/" target="_blank" title="Instagram"><i class="fab fa-instagram" style="color: #e1306c;"></i></a></div></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">Enjoy your stay!</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Contents</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#miniprot-%E5%9F%BA%E5%9B%A0%E6%B3%A8%E9%87%8A"><span class="toc-number">1.</span> <span class="toc-text">miniprot 基因注释</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6%E6%A0%BC%E5%BC%8F%E8%BD%AC%E6%8D%A2"><span class="toc-number">2.</span> <span class="toc-text">注释文件格式转换</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%B7%BB%E5%8A%A0%E5%BA%8F%E5%88%97%E4%B8%8E%E6%B3%A8%E9%87%8A%E6%96%87%E4%BB%B6"><span class="toc-number">3.</span> <span class="toc-text">添加序列与注释文件</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%BF%90%E8%A1%8C-SAW-makeRef"><span class="toc-number">4.</span> <span class="toc-text">运行 SAW makeRef</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E2%9A%A0%EF%B8%8F-%E6%B3%A8%E6%84%8F%E4%BA%8B%E9%A1%B9"><span class="toc-number">5.</span> <span class="toc-text">⚠️ 注意事项</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%8F%82%E8%80%83%E6%96%87%E7%8C%AE"><span class="toc-number">6.</span> <span class="toc-text">参考文献</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Posts</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2026/01/12/Lotka-Volterra_model/" title="动态系统稳定性与 Lotka-Volterra 模型浅略研究"><img src="https://oaknationalacademy-res.cloudinary.com/image/upload/v1700515161/xvjpxj8bmayxkhjawqkk.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="动态系统稳定性与 Lotka-Volterra 模型浅略研究"/></a><div class="content"><a class="title" href="/2026/01/12/Lotka-Volterra_model/" title="动态系统稳定性与 Lotka-Volterra 模型浅略研究">动态系统稳定性与 Lotka-Volterra 模型浅略研究</a><time datetime="2026-01-12T16:00:00.000Z" title="Created 2026-01-13 00:00:00">2026-01-13</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2026/01/11/SAW-makeRef/" title="构建 SAW 分析参考索引"><img src="https://oup.silverchair-cdn.com/oup/backfile/Content_public/Journal/bioinformatics/29/1/10.1093_bioinformatics_bts635/3/m_bioinformatics_29_1_15_f2.jpeg?Expires=1771021321&amp;Signature=fqt36-Ujsft5yhbtrBM0IVIOgNYIJdnH2nPfHkwb~JZD1RMAQ-Khdm6Qkf6X2m~MbpxQ4D~otvV37upgiKEy7dCA5r5wky4G9PL84IZxKK08t6oBzY1w-LB-gjzFS1gg4RMojlD~WYIgMjkBFkcAWdrrY3ZcoL1NwCkURaodOQbVmyn80P4SD89j3F9ylCvcKp6Dgn3Nj1yLcqgvxu9s-Rgs4cvbsJRWHSxHqpo0JmjtDDyFkzvDXHpwxsMaF60BvwMpvw5hMGBH0D2tBWm7Z~5jOkdpv0QySk~WoALpsNsgJndzXFF3jftYEyqN4h7fyDBN6DJREleM5lyhTwM2ow__&amp;Key-Pair-Id=APKAIE5G5CRDK6RD3PGA" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="构建 SAW 分析参考索引"/></a><div class="content"><a class="title" href="/2026/01/11/SAW-makeRef/" title="构建 SAW 分析参考索引">构建 SAW 分析参考索引</a><time datetime="2026-01-11T16:00:00.000Z" title="Created 2026-01-12 00:00:00">2026-01-12</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2026/01/05/hello-world/" title="Hello World"><img src="/figure/posts/hexo.png" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="Hello World"/></a><div class="content"><a class="title" href="/2026/01/05/hello-world/" title="Hello World">Hello World</a><time datetime="2026-01-05T16:00:00.000Z" title="Created 2026-01-06 00:00:00">2026-01-06</time></div></div></div></div></div></div></main><footer id="footer" style="background-image: url(https://www.vangoghgallery.com/img/starry_night_full.jpg);"><div class="footer-other"><div class="footer-copyright"><span class="copyright">&copy;&nbsp;2026 By Louis Xiong</span><span class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo 8.1.1</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly 5.5.3</a></span></div><div class="footer_custom_text">Thanks for visiting my blog!</div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Reading Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light and Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle Between Single-column and Double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside-config" type="button" title="Settings"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table of Contents"><i class="fas fa-list-ul"></i></button><button id="go-up" type="button" title="Back to Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js?v=5.5.3"></script><script src="/js/main.js?v=5.5.3"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui@6.1.7/dist/fancybox/fancybox.umd.min.js"></script><script src="https://cdn.jsdelivr.net/npm/instant.page@5.2.0/instantpage.min.js" type="module"></script><script src="https://cdn.jsdelivr.net/npm/node-snackbar@0.1.16/dist/snackbar.min.js"></script><div class="js-pjax"><script>(() => {
  const loadMathjax = () => {
    if (!window.MathJax) {
      window.MathJax = {
        loader: {
          load: [
            // Four font extension packages (optional)
            //- '[tex]/bbm',
            //- '[tex]/bboldx',
            //- '[tex]/dsfont',
            '[tex]/mhchem'
          ],
          paths: {
            'mathjax-newcm': '[mathjax]/../@mathjax/mathjax-newcm-font',

            //- // Four font extension packages (optional)
            //- 'mathjax-bbm-extension': '[mathjax]/../@mathjax/mathjax-bbm-font-extension',
            //- 'mathjax-bboldx-extension': '[mathjax]/../@mathjax/mathjax-bboldx-font-extension',
            //- 'mathjax-dsfont-extension': '[mathjax]/../@mathjax/mathjax-dsfont-font-extension',
            'mathjax-mhchem-extension': '[mathjax]/../@mathjax/mathjax-mhchem-font-extension'
          }
        },
        output: {
          font: 'mathjax-newcm',
        },
        tex: {
          inlineMath: [['$', '$'], ['\\(', '\\)']],
          tags: 'none',
          packages: {
            '[+]': [
              'mhchem'
            ]
          }
        },
        chtml: {
          scale: 1.1
        },
        options: {
          enableMenu: true,
          menuOptions: {
            settings: {
              enrich: false  // Turn off Braille and voice narration text automatic generation
            }
          },
          renderActions: {
            findScript: [10, doc => {
              for (const node of document.querySelectorAll('script[type^="math/tex"]')) {
                const display = !!node.type.match(/; *mode=display/)
                const math = new doc.options.MathItem(node.textContent, doc.inputJax[0], display)
                const text = document.createTextNode('')
                node.parentNode.replaceChild(text, node)
                math.start = {node: text, delim: '', n: 0}
                math.end = {node: text, delim: '', n: 0}
                doc.math.push(math)
              }
            }, '']
          }
        }
      }

      const script = document.createElement('script')
      script.src = 'https://cdn.jsdelivr.net/npm/mathjax@4.0.0/tex-mml-chtml.min.js'
      script.id = 'MathJax-script'
      script.async = true
      document.head.appendChild(script)
    } else {
      MathJax.startup.document.state(0)
      MathJax.texReset()
      MathJax.typesetPromise()
    }
  }

  btf.addGlobalFn('encrypt', loadMathjax, 'mathjax')
  window.pjax ? loadMathjax() : window.addEventListener('load', loadMathjax)
})()</script><script>(() => {
  const runMermaid = ele => {
    window.loadMermaid = true
    const theme = document.documentElement.getAttribute('data-theme') === 'dark' ? 'dark' : 'default'

    ele.forEach((item, index) => {
      const mermaidSrc = item.firstElementChild
      const config = mermaidSrc.dataset.config ? JSON.parse(mermaidSrc.dataset.config) : {}
      if (!config.theme) {
        config.theme = theme
      }
      const mermaidThemeConfig = `%%{init: ${JSON.stringify(config)}}%%\n`
      const mermaidID = `mermaid-${index}`
      const mermaidDefinition = mermaidThemeConfig + mermaidSrc.textContent

      const renderFn = mermaid.render(mermaidID, mermaidDefinition)
      const renderMermaid = svg => {
        mermaidSrc.insertAdjacentHTML('afterend', svg)
      }

      // mermaid v9 and v10 compatibility
      typeof renderFn === 'string' ? renderMermaid(renderFn) : renderFn.then(({ svg }) => renderMermaid(svg))
    })
  }

  const codeToMermaid = () => {
    const codeMermaidEle = document.querySelectorAll('pre > code.mermaid')
    if (codeMermaidEle.length === 0) return

    codeMermaidEle.forEach(ele => {
      const preEle = document.createElement('pre')
      preEle.className = 'mermaid-src'
      preEle.hidden = true
      preEle.textContent = ele.textContent
      const newEle = document.createElement('div')
      newEle.className = 'mermaid-wrap'
      newEle.appendChild(preEle)
      ele.parentNode.replaceWith(newEle)
    })
  }

  const loadMermaid = () => {
    if (true) codeToMermaid()
    const $mermaid = document.querySelectorAll('#article-container .mermaid-wrap')
    if ($mermaid.length === 0) return

    const runMermaidFn = () => runMermaid($mermaid)
    btf.addGlobalFn('themeChange', runMermaidFn, 'mermaid')
    window.loadMermaid ? runMermaidFn() : btf.getScript('https://cdn.jsdelivr.net/npm/mermaid@11.12.2/dist/mermaid.min.js').then(runMermaidFn)
  }

  btf.addGlobalFn('encrypt', loadMermaid, 'mermaid')
  window.pjax ? loadMermaid() : document.addEventListener('DOMContentLoaded', loadMermaid)
})()</script></div><canvas class="fireworks" mobile="true"></canvas><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/dist/fireworks.min.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/dist/activate-power-mode.min.js"></script><script>POWERMODE.colorful = true;
POWERMODE.shake = false;
POWERMODE.mobile = false;
document.body.addEventListener('input', POWERMODE);
</script><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/aplayer@1.10.1/dist/APlayer.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/aplayer@1.10.1/dist/APlayer.min.js"></script><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc@1.1.6/metingjs/dist/Meting.min.js"></script><script>(() => {
  const destroyAplayer = () => {
    if (window.aplayers) {
      for (let i = 0; i < window.aplayers.length; i++) {
        if (!window.aplayers[i].options.fixed) {
          window.aplayers[i].destroy()
        }
      }
    }
  }

  const runMetingJS = () => {
    typeof loadMeting === 'function' && document.getElementsByClassName('aplayer').length && loadMeting()
  }

  btf.addGlobalFn('pjaxSend', destroyAplayer, 'destroyAplayer')
  btf.addGlobalFn('pjaxComplete', loadMeting, 'runMetingJS')
})()</script><script src="/true" defer="defer"></script><script>document.addEventListener('DOMContentLoaded', () => {
  const pjaxSelectors = ["head > title","#config-diff","#body-wrap","#rightside-config-hide","#rightside-config-show",".js-pjax"]

  window.pjax = new Pjax({
    elements: 'a:not([target="_blank"])',
    selectors: pjaxSelectors,
    cacheBust: false,
    analytics: false,
    scrollRestoration: false
  })

  const triggerPjaxFn = (val) => {
    if (!val) return
    Object.values(val).forEach(fn => {
      try {
        fn()
      } catch (err) {
        console.debug('Pjax callback failed:', err)
      }
    })
  }

  document.addEventListener('pjax:send', () => {
    // removeEventListener
    btf.removeGlobalFnEvent('pjaxSendOnce')
    btf.removeGlobalFnEvent('themeChange')

    // reset readmode
    const $bodyClassList = document.body.classList
    if ($bodyClassList.contains('read-mode')) $bodyClassList.remove('read-mode')

    triggerPjaxFn(window.globalFn.pjaxSend)
  })

  document.addEventListener('pjax:complete', () => {
    btf.removeGlobalFnEvent('pjaxCompleteOnce')
    document.querySelectorAll('script[data-pjax]').forEach(item => {
      const newScript = document.createElement('script')
      const content = item.text || item.textContent || item.innerHTML || ""
      Array.from(item.attributes).forEach(attr => newScript.setAttribute(attr.name, attr.value))
      newScript.appendChild(document.createTextNode(content))
      item.parentNode.replaceChild(newScript, item)
    })

    triggerPjaxFn(window.globalFn.pjaxComplete)
  })

  document.addEventListener('pjax:error', e => {
    if (e.request.status === 404) {
      false
        ? pjax.loadUrl('/404.html')
        : window.location.href = e.request.responseURL
    }
  })
})</script><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><div id="local-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">Search</span><i class="fas fa-spinner fa-pulse" id="loading-status" hidden="hidden"></i><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="text-center" id="loading-database"><i class="fas fa-spinner fa-pulse"></i><span>  Loading Database</span></div><div class="local-search-input"><input placeholder="Find whatever you want..." type="text"/></div><hr/><div id="local-search-results"></div><div class="ais-Pagination" id="local-search-pagination" style="display:none;"><ul class="ais-Pagination-list"></ul></div><div id="local-search-stats"></div></div><div id="search-mask"></div><script src="/js/search/local-search.js?v=5.5.3"></script></div></div></body></html>